WIP: Improve break mesh by blocks

arovinelli opened the pull request:
This fix the debug mode tests for #12033

Commit:
all debug tests fails even ofr replicated #12033

roystgnr committed:
Fix node ids on ReplicatedMesh, unique ids

roystgnr committed:
Fixup id assignment, remove debugging output

This gets us to the point where id and unique_id values seem to all sync
up perfectly, but we're still getting mixed up boundary_id integers on
the new boundaries.

roystgnr committed:
Fix boundary ID assignment

Commit:
fix debug issue in #12033
still not Distributed

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/437761/) on e190465 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/14456/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/14456/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 3907f3aaa769ecdfa89be265b1b2d8808048b637`

moosebuild commented on large_media:
```diff
@@ -1 +1 @@
-Subproject commit e5f268b7624b4f08c47a02683607ac61b5f087da
+Subproject commit ee06a775fc6ef6ed7096392e355e18e0dadb23c2
```
**Caution!** This contains a submodule update

Commit:
fix large media

moosebuild commented on libmesh:
```diff
@@ -1 +1 @@
-Subproject commit 143b57032761c71a1df09c59ffc692bb4f198a0c
+Subproject commit 5db42ee857f6f5e1ab522c7f2ca8ae599f1e8b56
```
**Caution!** This contains a submodule update

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/437803/) on 48d68f7 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/14456/site)

This comment will be updated on new commits.

Commit:
all test should pass in normal and debug mode

Commit:
remove unused varaible

Commit:
debug distributed giving issues

arovinelli commented:
@roystgnr I don't want to annoy you too much with this, but there is something i can't really figure out. Most likely I'm missing something in the logic somewhere.
The new node id subdomain map is. However tehre is something odd because I always get this error in debug distributed mode which is coming from `void Elem::make_links_to_me_local(unsigned int n)`

it seems when it's trying to make a remote element local, it can't find the neighbdor but the connectivity and neighbor map is right and the same on each process.

```
nn = 4
neigh->n_sides() = 4
```

roystgnr commented:
Sorry, but what's the command line sequence I want to run here?  I tried
```
mpirun -np 4 ../../../moose_test-dbg --mesh-only -i generate_distributed_exodus.i
mpirun -np 4 ../../../moose_test-dbg --mesh-only -i break_mesh_2D_distributed.i
```
and I see a bunch of debugging output but no ...

Oh, wait, I have to add --distributed-mesh.  Yeah, I can replicate this.  I'll see if I can figure it out.

arovinelli commented:
> Sorry, but what's the command line sequence I want to run here? I tried
> 
> ```
> mpirun -np 4 ../../../moose_test-dbg --mesh-only -i generate_distributed_exodus.i
> mpirun -np 4 ../../../moose_test-dbg --mesh-only -i break_mesh_2D_distributed.i
> ```
> 
> and I see a bunch of debugging output but no ...
> 
> Oh, wait, I have to add --distributed-mesh. Yeah, I can replicate this. I'll see if I can figure it out.

@roystgnr Actually you just want to run the `break_mesh_2D_distributed.i` in distributed mode the other one is just generating an exodus with a mesh that might is set up to have at least a remote subdomain .
The actual contro sequence I used is 
`mpirun -np 4 ../../../moose_test-dbg -i break_mesh_2D_distributed.i --mesh-only --distributed-mesh --keep-cout --redirect-stdout`

the `--keep-cout` allows for std out from all ranks and `--redirect-stdout` generate a text file for each mpi process (but nothing is printed to screen), such that you don't go crazy while checking the output of a processes is correct.

roystgnr commented:
> the connectivity and neighbor map is right and the same on each process.

Doesn't look right to me.  On the failing process, elem 6 ("this") has neighbor link 0 pointing to elem 5 ("neigh"), when that neighbor link *should* be null, because the two points they share are now both duplicated nodes along the slit between them.  Looks like the other processors all have the same problem, but on the failing processor it's actually triggering during a mesh serialization because neigh->neighbor_ptr(2) is a remote_elem link rather than a link back to elem 6, so the neighbor links are actually inconsistent, not just wrong.

I'm guessing BreakMeshByBlockGenerator hasn't actually ever been breaking neighbor links when it splits apart two neighbors, but nothing's been really looking at the neighbor links before (at least not before ReplicatedMesh prepare_for_use could fix them) and so it hasn't caused problems until now?

arovinelli commented:
> > the connectivity and neighbor map is right and the same on each process.
> 
> Doesn't look right to me. On the failing process, elem 6 ("this") has neighbor link 0 pointing to elem 5 ("neigh"), when that neighbor link _should_ be null, because the two points they share are now both duplicated nodes along the slit between them. Looks like the other processors all have the same problem, but on the failing processor it's actually triggering during a mesh serialization because neigh->neighbor_ptr(2) is a remote_elem link rather than a link back to elem 6, so the neighbor links are actually inconsistent, not just wrong.

Just to clarify the aim of this is modifier is to split he mesh by adding nodes but leaving neighbor map intact, which means that two adjacent element with different block ids will not share the same nodes anymore, but they will still be neighbors. such that i can use a DG approach between them.
 
> I'm guessing BreakMeshByBlockGenerator hasn't actually ever been breaking neighbor links when it splits apart two neighbors, but nothing's been really looking at the neighbor links before (at least not before ReplicatedMesh prepare_for_use could fix them) and so it hasn't caused problems until now?

No because we want the links to remain intact.
Any idea on how to fix this?

roystgnr commented:
Oh, yikes.  We want to maintain neighbor links between elements that aren't technically neighbors?  We surely *can* support that in libMesh, but currently we *don't* - so to whatever extent it's been working in the past, that's been dumb luck.

The most likely culprit here might be a find_neighbors() call, then?  That typically strips down old neighbor links and then rebuilds them based on identifying patterns of shared nodes, and off the top of my head it's the only thing I can think of that isn't trying to respect existing neighbor links.

The failure is happening when a MeshSerializer gathers the mesh onto proc 0 for Exodus output.  Let me try looking at the state of things pre-gather and see if anything looks fishy.  It's also possible that the weird non-neighbor-neighbors situation here is a red herring and you've just managed to trip some unrelated corner case.

roystgnr commented:
Oh, hell, I get it now.  It's not an unrelated bug, but it's also not a failing in find_neighbors.

**Inside** make_links_to_me_local we're relying on a comparison between side_ptr results (shims to Side elements) to determine which of the neighbor's sides matches our side.  But if they're not true neighbors then *there is no match*.  If I receive a copy of previously-remote element A, and the information about that copy says its 3rd neighbor is element B, and element B currently has two remote_elem neighbors, then I have no way of knowing, aside from looking for matching nodes, which of those two remote_elem neighbors is supposed to be A!

I'm not sure what the right fix for this is.  Maybe communicate return neighbor pointer side indices for each neighbor when packing and unpacking an Elem?  That'd bump up communication size and code size, and I wouldn't dare commit it without doing a whole DistributedMesh sweep in CI, but I can't seem to think of any other option.

roystgnr commented:
Ooh, look.  We're comparing *centroids* of sides in BoundaryInfo::get_side_and_node_maps.  That might work here too.

roystgnr commented:
There's other places where we're doing direct side node comparison, though.  AbaqusIO, GmshIO, UNVIO, InfElemBuilder, MeshModification::all_tri, ReplicatedMesh::stitching_helper, find_neighbors of course... there may be other anticipated breakage in your future.

I think in *just* the Elem unpacking case it'll be safe enough and quick enough to compare centroids, though.  Let me try that.  Unless you guys ever combine fake neighbor pointers with sliver elements too?  In that case I think the safest thing for a nuclear design code to do would be ```libmesh_error("Just freeze in the dark, Bryukhanov")```.

arovinelli commented:
>  want to maintain neighbor links between elements that aren't technically neighbor?

Yes, one of the aim of `BreakMeshByblocks` is to allow cohesive zone modeling with a DG approach (breaking the mesh allow for a jump in the displacement field, but we need the neighbor information to compute it).

> If I receive a copy of previously-remote element A, and the information about that copy says its 3rd neighbor is element B, and element B currently has two remote_elem neighbors, then I have no way of knowing, aside from looking for matching nodes, which of those two remote_elem neighbors is supposed to be A! 
> I'm not sure what the right fix for this is. Maybe communicate return neighbor pointer side indices for each neighbor when packing and unpacking an Elem? That'd bump up communication size and code size, and I wouldn't dare commit it without doing a whole DistributedMesh sweep in CI, but I can't seem to think of any other option.

I see your point here, I don't really know how to help with this. If this is done only at the meshgeneration phase (which it doesn't seems the case) comparing node location or face centroids it's ok. but if it's done later (e.g. every time we dump files) then it will not work as the duplicated node might have different coordinates and face might be dispalced. (can you keep a copy of the undeformed mesh?) 

> I think in just the Elem unpacking case it'll be safe enough and quick enough to compare centroids, though. Let me try that. Unless you guys ever combine fake neighbor pointers with sliver elements too? In that case I think the safest thing for a nuclear design code to do would be libmesh_error("Just freeze in the dark, Bryukhanov").

I"m not sure what you mean by a `silver element` (an element with a very bad aspect ratio??). Why this should be a problem?
For the type of application we envisioned to use this method elements should all have nice geometries. In general we use

roystgnr commented:
Give https://github.com/libMesh/libmesh/pull/2362 a try?

roystgnr commented:
> then it will not work as the duplicated node might have different coordinates and face might be dispalced

Wait, are you planning to move these element pairs apart after displacement?  I threw in a tolerance for the floating point comparison to handle sides computing centroids in different orders, but that's going to work if the elements are O(epsilon) apart, not O(1).

> I"m not sure what you mean by a silver element (an element with a very bad aspect ratio??). Why this should be a problem?

If two faces of the same element collapse down onto each other than they can have the same centroid.

arovinelli commented:
> Oh, yikes. We want to maintain neighbor links between elements that aren't technically neighbors? We surely _can_ support that in libMesh, but currently we _don't_ - so to whatever extent it's been working in the past, that's been dumb luck.

By the way, all of this works great for a replicated mesh. I've done pretty big simulations with this. So it should have something to do with the serialization of the mesh when. Would using a nemesis output help?

roystgnr commented:
Not just serialization, but any redistribution.  So Nemesis would probably work (it doesn't need to gather any more ghost elements than we already have, right?) but any kind of repartitioning would fail in the same way.

Can you do refinement (even uniform) on the ReplicatedMesh case?  I'd expect find_neighbors to fail there.

arovinelli commented:
> Not just serialization, but any redistribution. So Nemesis would probably work (it doesn't need to gather any more ghost elements than we already have, right?) but any kind of repartitioning would fail in the same way.
> 
> Can you do refinement (even uniform) on the ReplicatedMesh case? I'd expect find_neighbors to fail there.

I was willing to try but the new meshgenerator system of moose doesn't seems to allow for uniform refinement in any meshgenerator block (@permcody ??) i could only use it with the mesh block but it conflicts with adding generators later. So i don't know  :sweat_smile:

Also @permcody can I use nemesis output for `--mesh-only`  executions?

permcody commented:
On Fri, Dec 6, 2019 at 12:05 PM Andrea Rovinelli <notifications@github.com>
wrote:

> Not just serialization, but any redistribution. So Nemesis would probably
> work (it doesn't need to gather any more ghost elements than we already
> have, right?) but any kind of repartitioning would fail in the same way.
>
> Can you do refinement (even uniform) on the ReplicatedMesh case? I'd
> expect find_neighbors to fail there.
>
> I was willing to try but the new meshgenerator system of moose doesn't
> seems to allow for uniform refinement in any meshgenerator block (
> @permcody <https://github.com/permcody> ??) i could only use it with the
> mesh block but it conflicts with adding generators later. So i don't know
> 😅
>

I believe somebody else already reported this. What I think we want to do
for generators is just have an explicit "uniform refine" generator that you
just insert into the generator tree. That way it's very flexible about when
it runs and we don't have to special case it.


> Also @permcody <https://github.com/permcody> can I use nemesis output for
> --mesh-only executions?
>

Should work fine.


> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/moose/pull/14456?email_source=notifications&email_token=AAXFOIBONEARRTSQGEMNZJDQXKPBHA5CNFSM4JSLLTPKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEGFBSTY#issuecomment-562698575>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAXFOIEBAMZJAF5IH72DZADQXKPBHANCNFSM4JSLLTPA>
> .
>

arovinelli commented:
@roystgnr thanks for finalising [libmesh/2363](https://github.com/libMesh/libmesh/pull/2363)
I think we still have some issue.  When I run `pirun -np 8 ../../../moose_test-dbg -i break_mesh_3D_polycrystal.i --distributed-mesh --mesh-only`

with the update lib mesh, I receive  the following errors 

```
Warning, ntags=3, but we currently only support reading 2 flags.
Assertion `mesh.comm().semiverify (elem ? &all_bcids : nullptr)' failed.

Assertion `mesh.comm().semiverify (elem ? &all_bcids : nullptr)' failed.

[Assertion `mesh.comm().semiverify (elem ? &all_bcids : nullptr)' failed.

[Assertion `mesh.comm().semiverify (elem ? &all_bcids : nullptr)' failed.

Assertion `mesh.comm().semiverify (elem ? &all_bcids : nullptr)' failed.

Assertion `mesh.comm().semiverify (elem ? &all_bcids : nullptr)' failed.

[Assertion `mesh.comm().semiverify (elem ? &all_bcids : nullptr)' failed.

[Assertion `mesh.comm().semiverify (elem ? &all_bcids : nullptr)' failed.

[7] /home/arovinelli/projects/NEW_stuff/moose/scripts/../libmesh/src/mesh/mesh_tools.C, line 1561, compiled Dec 12 2019 at 19:25:27
[0] /home/arovinelli/projects/NEW_stuff/moose/scripts/../libmesh/src/mesh/mesh_tools.C, line 1561, compiled Dec 12 2019 at 19:25:27
1] /home/arovinelli/projects/NEW_stuff/moose/scripts/../libmesh/src/mesh/mesh_tools.C, line 1561, compiled Dec 12 2019 at 19:25:27
2] /home/arovinelli/projects/NEW_stuff/moose/scripts/../libmesh/src/mesh/mesh_tools.C, line 1561, compiled Dec 12 2019 at 19:25:27
[3] /home/arovinelli/projects/NEW_stuff/moose/scripts/../libmesh/src/mesh/mesh_tools.C, line 1561, compiled Dec 12 2019 at 19:25:27
[4] /home/arovinelli/projects/NEW_stuff/moose/scripts/../libmesh/src/mesh/mesh_tools.C, line 1561, compiled Dec 12 2019 at 19:25:27
5] /home/arovinelli/projects/NEW_stuff/moose/scripts/../libmesh/src/mesh/mesh_tools.C, line 1561, compiled Dec 12 2019 at 19:25:27
6] /home/arovinelli/projects/NEW_stuff/moose/scripts/../libmesh/src/mesh/mesh_tools.C, line 1561, compiled Dec 12 2019 at 19:25:27
```

can you replicate this?

roystgnr commented:
This is incredibly frustrating:  **NO**, I can't replicate it this time!  I even tried a sweep over processor count, from 1 to 16, and everything seemed to run cleanly.

I'm using a ```--disable-strict-lgpl --enable-everything --enable-parmesh``` build, but I don't think the first two options should matter, and the third shouldn't matter in this case where there isn't a Mesh somewhere in the code path.

I'll try building on a different system; let me know what your configuration options are so I can match this as closely as possible?  That certainly *looks* like a real error, boundary conditions that aren't correctly matching between local and ghosted copies of the same element.

permcody commented:
Andrea, per our conversation offline, can you replicate this in your "DEER"
app? If we can replicate this on our build boxes, it'll help to figure out
your issue.

On Fri, Dec 13, 2019 at 9:03 AM roystgnr <notifications@github.com> wrote:

> This is incredibly frustrating: *NO*, I can't replicate it this time! I
> even tried a sweep over processor count, from 1 to 16, and everything
> seemed to run cleanly.
>
> I'm using a --disable-strict-lgpl --enable-everything --enable-parmesh
> build, but I don't think the first two options should matter, and the third
> shouldn't matter in this case where there isn't a Mesh somewhere in the
> code path.
>
> I'll try building on a different system; let me know what your
> configuration options are so I can match this as closely as possible? That
> certainly *looks* like a real error, boundary conditions that aren't
> correctly matching between local and ghosted copies of the same element.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/moose/pull/14456?email_source=notifications&email_token=AAXFOIBELOXX7BFLA2FUOWDQYOW4VA5CNFSM4JSLLTPKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEG2NAZY#issuecomment-565497959>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAXFOICR5O6HSHTIZBHRDP3QYOW4VANCNFSM4JSLLTPA>
> .
>

Commit:
remove forbidden words

moosebuild commented on libmesh:
```diff
@@ -1 +1 @@
-Subproject commit 83ce83b325b7e1e0e4e0ea1b426b4566a3d5a4f5
+Subproject commit 22ef4128edc89f2e61de767f0a54afec25360aef
```
**Caution!** This contains a submodule update

arovinelli replied:
this is required to check everything works, hopefully i update libmesh correctly

arovinelli commented:
> Andrea, per our conversation offline, can you replicate this in your "DEER" app? If we can replicate this on our build boxes, it'll help to figure out your issue.
> […](#)
> On Fri, Dec 13, 2019 at 9:03 AM roystgnr ***@***.***> wrote: This is incredibly frustrating: *NO*, I can't replicate it this time! I even tried a sweep over processor count, from 1 to 16, and everything seemed to run cleanly. I'm using a --disable-strict-lgpl --enable-everything --enable-parmesh build, but I don't think the first two options should matter, and the third shouldn't matter in this case where there isn't a Mesh somewhere in the code path. I'll try building on a different system; let me know what your configuration options are so I can match this as closely as possible? That certainly *looks* like a real error, boundary conditions that aren't correctly matching between local and ghosted copies of the same element. — You are receiving this because you were mentioned. Reply to this email directly, view it on GitHub <#14456?email_source=notifications&email_token=AAXFOIBELOXX7BFLA2FUOWDQYOW4VA5CNFSM4JSLLTPKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEG2NAZY#issuecomment-565497959>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AAXFOICR5O6HSHTIZBHRDP3QYOW4VANCNFSM4JSLLTPA> .

arovinelli commented:
@permcody  and @roystgnr I pushed the branch hear and hopefully I updated libmesh correctly. Let's see what the testing are the results of the tests I susepct I'm doing something wrong with libmesh updates in my machine

moosebuild commented on libmesh:
```diff
@@ -1 +1 @@
-Subproject commit 83ce83b325b7e1e0e4e0ea1b426b4566a3d5a4f5
+Subproject commit 22ef4128edc89f2e61de767f0a54afec25360aef
```
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

Commit:
disable some tests depending on breakmeshgenerator

Commit:
this needs a lot of cleaning but should pass all tests

arovinelli commented:
@roystgnr and @permcody I did some intensive teting and I believe the problem I reported before was a matter of correctly updating libmesh and moose (the libmesh PR royworked on should suffice to fix the issue). The failure that I see now in the tests instead seems to be related to a mistake in my updated code logic that messes up boundaries creation. I'll work on that.
However i want to mention that exodiff cannot be used to check if files are correct for this generator, because when comampring nodes, it doesn't check the node element connectivity but only the node position. So it's impossible to say from an exodiff if everything is correct. Furthemore, exodiff doesn't check for boudanries

Commit:
let's see waht is failing now

arovinelli commented:
@roystgnr and @permcody to me this should work, but apparently I'm missing something.
All distributed test involving an `InterfaceKernel` and or an `InterfaceMaterial` fail with the following error
```
Assertion `!node_touched_by_anyone[nodeid] || node_touched_by_me[nodeid]' failed.
[0] /home/arovinelli/projects/NEW_stuff/moose/scripts/../libmesh/src/mesh/mesh_tools.C, line 1866, compiled Dec 13 2019 at 12:35:04

```

However the `BreakMeshByBlockGenerator` pass all tests with the `--mesh-only` flag. I'm a bit puzzled now

lindsayad commented:
What's the full backtrace?

roystgnr commented:
Seconded on the backtrace question.

What's happening on that assertion failure is that libMesh is trying to make sure that all our nodes are owned by a processor which also owns an element connected to that node, and we've found a node whose processor_id() doesn't obey that requirement.  Normally this is an internal failure, because during refinement/repartitioning/etc we automatically repartition nodes as necessary to ensure that "topological consistency" is maintained, but depending on what code paths you're using after creating your new nodes perhaps you're not reaching that repartition?

Sorry for the stringency of that requirement, but there are at least one or two DistributedMesh algorithms that depend on being able to reach any local node via a local element, and that IIRC would either be more complicated or less efficient if we couldn't guarantee that.

arovinelli commented:
@lindsayad How do i get the backtrace in parallel?

@roystgnr I believe what you are saying is true, and it might be the cause of failure. I'm assigning the new node to the cpu owning the original node. for what you said I should assign the node to the cpu owning an element connected to that node. Satisfying these requiriment is tricky because I need to decide the cpu id a priori to satisfy all cases (think about complicated tetmeshes). @roystgnr If I understand it correctly the requirements are:

1.  the new node must be assigned to any cpu owning an element connected to that node
2. for the case in which the new node is added by multiple partitions I need to assign its cpu consistently (e.g. to the same cpu)

roystgnr commented:
> How do i get the backtrace in parallel?

Just get it on proc 0; this is an internal parallel_only() function so every other processor on the communicator ought to be in the same place.

roystgnr commented:
1 is correct.  We *used* to demand that the node be owned by the lowest-ranked processor owning a connected element, but fortunately that load-imbalanced rule should no longer be a requirement.

2 is technically true, and getting that consistent yourself is the best way to do things, but if for some reason you can't get proc ids consistent on the first pass, there are a few mesh_communication.h and parallel_ghost_sync.h methods that can fix it for you.  If you need that then let me know and I'll go over the requirements-vs-costs of the various options.

arovinelli commented:
> > How do i get the backtrace in parallel?
> 
> Just get it on proc 0; this is an internal parallel_only() function so every other processor on the communicator ought to be in the same place.

I don't know how to do this on gdb in linux, a command line example will really help

arovinelli commented:
> 1 is correct. We _used_ to demand that the node be owned by the lowest-ranked processor owning a connected element, but fortunately that load-imbalanced rule should no longer be a requirement.

this is what I was thinking to do (easiest way)


> 2 is technically true, and getting that consistent yourself is the best way to do things, but if for some reason you can't get proc ids consistent on the first pass, there are a few mesh_communication.h and parallel_ghost_sync.h methods that can fix it for you. If you need that then let me know and I'll go over the requirements-vs-costs of the various options.

Doing it manually would involve some other communication but it shouldn't be that bad

roystgnr commented:
> I don't know how to do this on gdb in linux, a command line example will really help

I'm guessing you didn't configure with --enable-tracefiles or --enable-everything?  (if you have, just check the traceout files)

My personal favorite hack for small runs is my ```mpirunxterm``` script:
```
#!/bin/bash
  
numproc=$2

shift
shift

exec mpirun --oversubscribe -np $numproc xterm -e bash -c "$* && read -p COMPLETED finalinput || read -p FAILED finalinput; sleep 20"
```
which combines with my .gdbrc settings:
```
set breakpoint pending on
set pagination off
b MPI_Abort
b libMesh::MacroFunctions::report_error
```
to let me run things like:
```
mpirunxterm -np 4 gdb -ex r --args ./my_application --my_args
```
or
```
LIBMESH_RUN='mpirunxterm -np 4 libtool --mode=execute gdb -ex r --args' make check -C tests/
```

lindsayad commented:
In MOOSE you can also do `mpiexec -np 2 ./exe-dbg -i input.i
--start-in-debugger='gdb'`

On Wed, Dec 18, 2019 at 12:05 PM roystgnr <notifications@github.com> wrote:

> I don't know how to do this on gdb in linux, a command line example will
> really help
>
> I'm guessing you didn't configure with --enable-tracefiles or
> --enable-everything? (if you have, just check the traceout files)
>
> My personal favorite hack for small runs is my mpirunxterm script:
>
> #!/bin/bash
>
> numproc=$2
>
> shift
> shift
>
> exec mpirun --oversubscribe -np $numproc xterm -e bash -c "$* && read -p COMPLETED finalinput || read -p FAILED finalinput; sleep 20"
>
> which combines with my .gdbrc settings:
>
> set breakpoint pending on
> set pagination off
> b MPI_Abort
> b libMesh::MacroFunctions::report_error
>
> to let me run things like:
>
> mpirunxterm -np 4 gdb -ex r --args ./my_application --my_args
>
> or
>
> LIBMESH_RUN='mpirunxterm -np 4 libtool --mode=execute gdb -ex r --args' make check -C tests/
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/moose/pull/14456?email_source=notifications&email_token=ACOGA4E3NKBF7ZG2DPHMVLLQZJX6JA5CNFSM4JSLLTPKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEHHESAQ#issuecomment-567167234>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACOGA4EQK46A3S4UOAPOU3LQZJX6JANCNFSM4JSLLTPA>
> .
>

arovinelli commented:
@lindsayad and @roystgnr thanks for the tips.
Here is the stacktrace of rank 0

```
(gdb) #1  0x00007fe3876f6c94 in libMesh::MeshTools::libmesh_assert_topology_consistent_procids<libMesh::Node> (mesh=...)
(gdb)     at /home/arovinelli/projects/NEW_stuff/moose/scripts/../libmesh/src/mesh/mesh_tools.C:1865
A program is being debugged already.  Kill it? (y or n) EOF [assumed Y]
Illegal process-id: /home/arovinelli/projects/NEW_stuff/moose/scripts/../libmesh/src/mesh/mesh_tools.C:1865.
(gdb) #2  0x00007fe386ddbfd8 in libMesh::MeshTools::libmesh_assert_valid_procids<libMesh::Node> (mesh=...) at ./include/libmesh/mesh_tools.h:583
(gdb) #3  0x00007fe386dba061 in libMesh::DofMap::distribute_local_dofs_var_major (
(gdb)     this=0x1de7790, next_free_dof=@0x7ffda1f9b3b4: 9, mesh=...)
Undefined command: "this".  Try "help".
(gdb)     at /home/arovinelli/projects/NEW_stuff/moose/scripts/../libmesh/src/base/dof_map.C:1424
Illegal process-id: /home/arovinelli/projects/NEW_stuff/moose/scripts/../libmesh/src/base/dof_map.C:1424.
(gdb) #4  0x00007fe386db63e6 in libMesh::DofMap::distribute_dofs (this=0x1de7790, 
(gdb)     mesh=...)
Undefined command: "mesh".  Try "help".
(gdb)     at /home/arovinelli/projects/NEW_stuff/moose/scripts/../libmesh/src/base/dof_map.C:951
Illegal process-id: /home/arovinelli/projects/NEW_stuff/moose/scripts/../libmesh/src/base/dof_map.C:951.
(gdb) #5  0x00007fe387b155b0 in libMesh::System::init_data (this=0x1de6c90)
(gdb)     at /home/arovinelli/projects/NEW_stuff/moose/scripts/../libmesh/src/systems/system.C:271
Illegal process-id: /home/arovinelli/projects/NEW_stuff/moose/scripts/../libmesh/src/systems/system.C:271.
(gdb) #6  0x00007fe387af8ebf in libMesh::ImplicitSystem::init_data (this=0x1de6c90)
(gdb)     at /home/arovinelli/projects/NEW_stuff/moose/scripts/../libmesh/src/systems/implicit_system.C:93
Illegal process-id: /home/arovinelli/projects/NEW_stuff/moose/scripts/../libmesh/src/systems/implicit_system.C:93.
(gdb) #7  0x00007fe387b15445 in libMesh::System::init (this=0x1de6c90)
(gdb)     at /home/arovinelli/projects/NEW_stuff/moose/scripts/../libmesh/src/systems/---Type <return> to continue, or q <return> to quit---\
system.C:249
Illegal process-id: /home/arovinelli/projects/NEW_stuff/moose/scripts/../libmesh/src/systems/---Type <return> to continue, or q <return> to quit---system.C:249.
(gdb) #8  0x00007fe387ab912f in libMesh::EquationSystems::init (this=0x1db4988)
(gdb)     at /home/arovinelli/projects/NEW_stuff/moose/scripts/../libmesh/src/systems/equation_systems.C:111
Illegal process-id: /home/arovinelli/projects/NEW_stuff/moose/scripts/../libmesh/src/systems/equation_systems.C:111.
(gdb) #9  0x00007fe38c855417 in FEProblemBase::init (this=0x1db3fc8)
(gdb)     at /home/arovinelli/projects/NEW_stuff/moose/framework/src/problems/FEProblemBase.C:4588
Illegal process-id: /home/arovinelli/projects/NEW_stuff/moose/framework/src/problems/FEProblemBase.C:4588.
(gdb) #10 0x00007fe38c7868d1 in InitProblemAction::act (this=0x1b18668)
(gdb)     at /home/arovinelli/projects/NEW_stuff/moose/framework/src/actions/InitProblemAction.C:30
Illegal process-id: /home/arovinelli/projects/NEW_stuff/moose/framework/src/actions/InitProblemAction.C:30.
(gdb) #11 0x00007fe38c7648d6 in Action::timedAct (this=0x1b18668)
(gdb)     at /home/arovinelli/projects/NEW_stuff/moose/framework/src/actions/Action.C:93
Illegal process-id: /home/arovinelli/projects/NEW_stuff/moose/framework/src/actions/Action.C:93.
(gdb) #12 0x00007fe38c76a144 in ActionWarehouse::executeActionsWithAction (
(gdb)     this=0x1b21160, task="init_problem")
Undefined command: "this".  Try "help".
(gdb)     at /home/arovinelli/projects/NEW_stuff/moose/framework/src/actions/ActionWarehouse.C:364
Illegal process-id: /home/arovinelli/projects/NEW_stuff/moose/framework/src/actions/ActionWarehouse.C:364.
(gdb) #13 0x00007fe38c769cfd in ActionWarehouse::executeAllActions (this=0x1b21160)
(gdb)     at /home/arovinelli/projects/NEW_stuff/moose/framework/src/actions/ActionWarehouse.C:338
Illegal process-id: /home/arovinelli/projects/NEW_stuff/moose/framework/src/actions/ActionWarehouse.C:338.
(gdb) #14 0x00007fe38d17880a in MooseApp::runInputFile (this=0x1b200f8)
(gdb)     at /home/arovinelli/projects/NEW_stuff/moose/framework/src/base/MooseApp.C:871
Illegal process-id: /home/arovinelli/projects/NEW_stuff/moose/framework/src/base/MooseApp.C:871.
(gdb) (gdb) \#0  0x00007fe380773920 in __cxa_throw () from /lib64/libstdc++.so.6
```

Commit:
ideally, node touched error should be fixed

arovinelli commented:
@permcody @lindsayad  @roystgnr after fixing the node_processor_id, this pr passes all the tests for distributed mesh, however I still have an issue with `interface_material` doesn't seem to work properly with a distributed mesh. Might this a relationship manager problem?

roystgnr commented:
> I still have an issue with interface_material doesn't seem to work properly with a distributed mesh

Could you be more specific?

arovinelli commented:
> > I still have an issue with interface_material doesn't seem to work properly with a distributed mesh
> 
> Could you be more specific?
@roystgn I was wrong it wtill hit this exception `Assertion `!node_touched_by_anyone[nodeid] || node_touched_by_me[nodeid]' failed.`
Could you please tell what should I do to let libmesh figure out the process id for new nodes by itself? I'm starting to be annoyed by this

Commit:
update some Moose::out

arovinelli commented:
@roystgnr so for the failing cases in the [distributed mesh tests](https://civet.inl.gov/job/449308/) I'm sure that the new node ifs and node  processor ids are correct. However I still hit the assertion and I do not understand why. The new nodes are going to be added to subdomain zeros and elements 1 and 3. Both elements 1 and 3 are on cpu 0 and therefore I assign the new nodes (id 9 10 11) processor ids to  cpu 0.  The element node map and node cpu map area consistent across all cpus. I do not understand why this is happening.
YYou should be able to replicate this by going to `moose/test/tests/materials/interface_material` and exectuing `mpiexec -n 2 ../../../moose_test-dbg -i interface_value_material_split_mesh.i --keep-cout --redirect-stdout`

Elem_node_final:
```
RANK 0 
OUT_ELEMENT_NODE_START
0 0 1 2 3 
1 9 4 5 10 
2 3 2 6 7 
3 10 5 8 11 
OUT_ELEMENT_NODE_END

RANK1
OUT_ELEMENT_NODE_START
0 0 1 2 3 
1 9 4 5 10 
2 3 2 6 7 
3 10 5 8 11 
```

Node processor_id
```
RANK 0 
OUT_NODE_PID_START
0 1
1 0
2 0
3 1
4 0
5 0
6 0
7 1
8 0
9 0
10 0
11 0
OUT_NODE_PID_END

RANK1
OUT_NODE_PID_START
0 1
1 0
2 0
3 1
4 0
5 0
6 0
7 1
8 0
9 0
10 0
11 0
OUT_NODE_PID_END

```

arovinelli commented:
@roystgnr any idea on what I could be missing that could cause the assertion to fail?

Is [this](https://github.com/arovinelli/moose/blob/e6abe125a14d55e3d269f57a929ad627be9b76ac/framework/src/meshgenerators/BreakMeshByBlockGenerator.C#L205-L213)  enough when adding new nodes and assigning new processor_ids or I'm missing something?

lindsayad commented:
I have not looked at this at all, but it's possible that this is a problem on the MOOSE side with `InterfaceMaterial`

Commit:
added test to interfacekernel

arovinelli commented:
> I have not looked at this at all, but it's possible that this is a problem on the MOOSE side with `InterfaceMaterial`

@lindsayad I tried playing a bit with this, excluding the `interface_material` and just using an `interface_kernel` (see `test/tests/interfacekernels/2d_interface/coupled_value_coupled_flux_split_mesh.i`) what it comes out is that different cpus sees a different number of nodes (see output below). So I believe I am missing something in the modifier (like set maximum number of nodes or something like that [https://github.com/arovinelli/moose/blob/e6abe125a14d55e3d269f57a929ad627be9b76ac/framework/src/meshgenerators/BreakMeshByBlockGenerator.C#L205-L213](https://github.com/arovinelli/moose/blob/e6abe125a14d55e3d269f57a929ad627be9b76ac/framework/src/meshgenerators/BreakMeshByBlockGenerator.C#L205-L213) )


When I'm adding the node in the distri

```
RANK 0
Parallelism:
  Num Processors:          2
  Num Threads:             1

Mesh: 
  Parallel Type:           distributed
  Mesh Dimension:          2
  Spatial Dimension:       2
    Nodes:                  
    Total:                 9
    Local:                 6
  Elems:                   
    Total:                 4
    Local:                 2
  Num Subdomains:          2
  Num Partitions:          2
  Partitioner:             parmetis

RANK1 
Parallelism:
  Num Processors:          2
  Num Threads:             1

Mesh: 
  Parallel Type:           distributed
  Mesh Dimension:          2
  Spatial Dimension:       2
    Nodes:                   
    Total:                 9
    Local:                 3
  Elems:                   
    Total:                 4
    Local:                 2
  Num Subdomains:          2
  Num Partitions:          2
  Partitioner:             parmetis
```

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/449937/) on 6fd71fe wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/14456/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/14456/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 6117f17d01f291d21fbd86ee04f1af89c3eb3429`

Commit:
fixed test file

Commit:
modified interfacekernel test

lindsayad commented:
I think I'm going to try and revive this so that I can  create  #14905 without any  `METHOD` or `mesh_type` restrictions (maybe I'll ignore `recover` for now). It seems like this was getting close

lindsayad commented:
@roystgnr Are you not supposed to call `add_node` for ghost nodes? I'm kind of flummoxed by this logic in `DistributedMesh::add_node`:
```c++
  // Try to make the cached node data more accurate
  if (node_procid == this->processor_id() ||
      node_procid == DofObject::invalid_processor_id)
    _n_nodes++;
```
Why don't we increment for every node? If I run the `moose/test/tests/interfacekernels/ik_displaced/different_jxw_displaced.i` test with two procs in distributed mode, I fail this assertion in `DistributedMesh::delete_remote_elements`:  `libmesh_assert_equal_to (this->n_nodes(), this->parallel_n_nodes());` because `_n_nodes` isn't incremented when I add a ghost node.

With the current code @arovinelli has, he has guaranteed (as far as I can tell) that he will be adding nodes with consistent ids and consistent proc_ids across processes. Shouldn't this work? Does have have to set the node processor_id to `invalid_processor_id` for any node for which `node->processor_id != this->processor_id()`? This seems silly since he's already decided the partitioning up front.

lindsayad commented:
I'm thinking we must not be calling `prepare_for_use` or more particularly `update_parallel_id_counts` after this mesh generator is run and before `delete_remote_elements`...In that case I suppose this is our (MOOSE's) fault...

arovinelli commented:
> I'm thinking we must not be calling `prepare_for_use` or more particularly `update_parallel_id_counts` after this mesh generator is run and before `delete_remote_elements`...In that case I suppose this is our (MOOSE's) fault...

@lindsayad and @roystgnr 
I don't know about this specifically, but I remember having troubles with this when assigning `invalid_processor_id` `invalid_dof_id` and let libmesh figuring these out for me. At some point I concluded that there was an issue if new nodes are coincident with other nodes. Unfortunately I don't have a solid proof, but it might be a clue

lindsayad committed:
prepare_for_use in BreakMeshByBlockGenerator

There's no guarantee that we're going to prepare later. During
construction, MooseMesh._needs_prepare_for_use is initialized to
`false`; I think this is a fair approach...if at any point during the
MeshGeneration phase someone does something that needs mesh preparation,
they should call `prepare_for_use` on the spot. In that vein, we call
`prepare_for_use` here because we've done an operation that requires
updating the parallel counts.

lindsayad commented:
@permcody We don't ever run parallel distributed debug moose test do we? (I don't think we actually run parallel debug moose test). The mesh generator results aren't pretty

arovinelli commented:
@lindsayad I’m pretty sure not. At that time I had to explicitly force parallel debug distributed mesh tests for this  PR.

Sent from my iPhone

On Mar 18, 2020, at 4:20 PM, Alex Lindsay <notifications@github.com> wrote:

﻿

@permcody<https://github.com/permcody> We don't ever run parallel distributed debug moose test do we? (I don't think we actually run parallel debug moose test). The mesh generator results aren't pretty

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/idaholab/moose/pull/14456#issuecomment-600864621>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AIIMUU3ZKMVUGCOH4SVQIRLRIE3KVANCNFSM4JSLLTPA>.

permcody commented:
parallel distributed debug - that's a mouthful and nope, we don't yet.

permcody commented:
However, if we need to, I suppose we could - what a nightmare.

lindsayad commented:
> There's other places where we're doing direct side node comparison, though. AbaqusIO, GmshIO, UNVIO, InfElemBuilder, MeshModification::all_tri, ReplicatedMesh::stitching_helper, find_neighbors of course... there may be other anticipated breakage in your future.

Indeed my last commit https://github.com/idaholab/moose/pull/14456/commits/876e0b37d57bd6a6112f08e7b8af6b410b7ef132 broke everything just by introducing a `prepare_for_use` which unwittingly called `find_neighbors`. If I call `prepare_for_use` with `skip_find_neighbors`, then the `moose/test/tests/meshgenerators/break_mesh_by_block_generator/break_mesh_3D_auto.i` input runs like a dream. This is also a testament to how in MOOSE instead of calling `prepare_for_use` all the time, we actually hardly ever `prepare_for_use` at all.

Anyway after breaking by block, `find_neighbors` wipes out the previous neighbors and then doesn't add back neighbors which are on a different subdomain than `elem` because `Elem==` does direct node comparison as @roystgnr stated. With those different-subdomain neighbor links wiped out, our ghosting functors won't try to keep them around and they will get deleted when running in a distributed mode.

Relying on never calling `prepare_for_use` without `skip_find_neighbors=true` is not a viable option. We have no idea what mesh generators down the pipe might do (or maybe crazily we might actually call `prepare_for_use` after *all* mesh generators have been executed sometime in the future). I'm going to have to go through the huge discussion surrounding this topic again, but @roystgnr is it too dangerous to *not* do node  comparisons in `Elem==` and instead do centroid comparisons? I'm guessing the concern with the latter might be false positives with things like sliver elements etc.

roystgnr commented:
Man, even mid-plague I can't take three days' vacation, huh?

> Are you not supposed to call add_node for ghost nodes?

You are, but only if you're manually handling the node id too.

> Why don't we increment for every node?

Good question, but the problem is it shouldn't be an important one: in theory, in mid-mesh-modification like that, the cached _n_nodes should just be used for preallocation and other performance issues, not as anything you can rely on, because other partitions might be adding nodes that you can't see or count.

> I fail this assertion in DistributedMesh::delete_remote_elements: libmesh_assert_equal_to (this->n_nodes(), this->parallel_n_nodes()); because _n_nodes isn't incremented when I add a ghost node.

No, it's because you didn't ```update_parallel_id_counts()``` after adding all your ghost nodes.  Even if we incremented for every add_node, you'd still fail that assertion as soon as any processor had a remote node, if you're not updating.

> @roystgnr is it too dangerous to not do node comparisons in Elem== and instead do centroid comparisons? 

I know I'm slow on the draw on this one, but yes, very.

> Relying on never calling prepare_for_use without skip_find_neighbors=true is not a viable option.

No, but it's the best we can do for now.  Relying on maintaining neighbor links between non-neighbors was never a great option, IMHO.  We might as well tweak a std::set's underlying red-black tree sort and complain that some set algorithms stop working down the line when they get a non-sorted set.

The long-term solution might be to completely rip out find_neighbors, and instead update neighbor pointers solely on the fly as mesh modifications are made.  But that's something I've been avoiding because I'm expecting it to be a major sink of man-hours; I'd rather not bite that bullet until we've got no other choice.

lindsayad commented:
> No, but it's the best we can do for now. Relying on maintaining neighbor links between non-neighbors was never a great option, IMHO.

With respect to the reference mesh, even though we've "broken" the geometry and inserted new nodes, I think one could argue that conceptually and geometrically the now "broken" pairs are still "neighbors". What about per https://github.com/libMesh/libmesh/pull/2474#issuecomment-601252223 in some `Elem::is_coincident` method comparing node *coordinates* instead of node ids?

roystgnr commented:
> I think one could argue that conceptually and geometrically the now "broken" pairs are still "neighbors".

Oh, definitely, but you could say the same about periodic boundary condition neighbors in many problems, and yet we weren't safe using the neighbor links for those; we had to add new APIs.  I'd love to be able to do arbitrary quotient topologies, there's just a lot of obstacles between here and there.

> What about per libMesh/libmesh#2474 (comment) in some Elem::is_coincident method comparing node coordinates instead of node ids?

Nope; IIRC that's actually what we did for a span of several years (due to a *different* technical limitation), and it was a big advance to switch to ids because that's how we now manage to support slit meshes without accidentally stitching them back together again.

I'm not sure what we can do as a short term fix other than avoiding the find_neighbors() call.

lindsayad commented:
> I'm not sure what we can do as a short term fix other than avoiding the find_neighbors() call.

Alright, fair enough. It's not satisfying, but it's also the least amount of work...

lindsayad committed:
Skip find neighbors

lindsayad commented:
This  PR will have to wait until something like libmesh/libmesh#2473 goes in. As outlined in lindsayad/libmesh@48e4c8c in current libmesh `master` whenever we serialize the mesh for exodus output (which we do all the time even when running with distributed mesh...yes, yes people should be using nemesis), we call `UnstructuredMesh::find_neighbors` which destroys the neighbor links. This destroying of neighbor links typically happens as early as initial output, which means that by the time we actually run the simulation solve the neighbor links are gone, we have a bunch of `nullptrs` instead of neighbors we want, and we're typically not going to even call interface kernel residual and jacobian methods because of the latter fact

lindsayad commented:
Nope

> On Apr 24, 2020, at 10:14 PM, stale[bot] <notifications@github.com> wrote:
> 
> ﻿
> This pull request has been automatically marked as stale because it has not had recent activity. It will be closed if no further activity occurs. Thank you for your contributions.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.

lindsayad commented:
Don’t do it

> On May 25, 2020, at 12:08 AM, stale[bot] <notifications@github.com> wrote:
> 
> ﻿
> This pull request has been automatically marked as stale because it has not had recent activity. It will be closed if no further activity occurs. Thank you for your contributions.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.

lindsayad commented:
I still want to make this work, especially given recent discussions about
ordering of mesh generator execution and ghosting functor addition

On Thu, Jun 25, 2020 at 8:50 PM stale[bot] <notifications@github.com> wrote:

> This pull request has been automatically marked as stale because it has
> not had recent activity. It will be closed if no further activity occurs.
> Thank you for your contributions.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/moose/pull/14456#issuecomment-649943847>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACOGA4FWNOTVT7CQMZIOLC3RYQLJ7ANCNFSM4JSLLTPA>
> .
>
